sae_training/train_sae_on_language_model.py

```diff
@@ -95,240 +258,195 @@ def train_sae_on_language_model(
-                            f"sparsity/dead_features{suffix}": ghost_grad_neuron_mask.sum().item(),
-                            f"details/current_learning_rate{suffix}": current_learning_rate,
-                            "details/n_training_tokens": n_training_tokens,
-                        },
-                        step=n_training_steps,
-                    )
-
-                # record loss frequently, but not all the time.
-                if use_wandb and (
-                    (n_training_steps + 1) % (wandb_log_frequency * 10) == 0
-                ):
-                    sparse_autoencoder.eval()
-                    suffix = wandb_log_suffix(sae_group.cfg, hyperparams)
-                    run_evals(
-                        sparse_autoencoder,
-                        activation_store,
-                        model,
-                        n_training_steps,
-                        suffix=suffix,
-                    )
-                    sparse_autoencoder.train()
-
-            loss.backward()
-            sparse_autoencoder.remove_gradient_parallel_to_decoder_directions()
-            optimizer[i].step()
 
-        # checkpoint if at checkpoint frequency
-        if n_checkpoints > 0 and n_training_tokens > checkpoint_thresholds[0]:
-            path = f"{sae_group.cfg.checkpoint_path}/{n_training_tokens}_{sae_group.get_name()}.pt"
-            for sae in sae_group:
-                sae.set_decoder_norm_to_unit_norm()
-            sae_group.save_model(path)
-
-            log_feature_sparsity_path = f"{sae_group.cfg.checkpoint_path}/{n_training_tokens}_{sae_group.get_name()}_log_feature_sparsity.pt"
-            log_feature_sparsity = []
-            for sae_id in range(len(sae_group)):
-                feature_sparsity = (
-                    act_freq_scores[sae_id] / n_frac_active_tokens[sae_id]
-                )
-                log_feature_sparsity.append(
-                    torch.log10(feature_sparsity + 1e-10).detach().cpu()
-                )
-            torch.save(log_feature_sparsity, log_feature_sparsity_path)
+        ctx.act_freq_scores = torch.zeros(
+            sparse_autoencoder.cfg.d_sae, device=sparse_autoencoder.cfg.device
+        )
+        ctx.n_frac_active_tokens = 0
+
+    ghost_grad_neuron_mask = (
+        ctx.n_forward_passes_since_fired > sparse_autoencoder.cfg.dead_feature_window
+    ).bool()
+
+    # Forward and Backward Passes
+    (
+        sae_out,
+        feature_acts,
+        loss,
+        mse_loss,
+        l1_loss,
+        ghost_grad_loss,
+    ) = sparse_autoencoder(
+        sae_in,
+        ghost_grad_neuron_mask,
+    )
+    did_fire = (feature_acts > 0).float().sum(-2) > 0
+    ctx.n_forward_passes_since_fired += 1
+    ctx.n_forward_passes_since_fired[did_fire] = 0
+
+    with torch.no_grad():
+        # Calculate the sparsities, and add it to a list, calculate sparsity metrics
+        ctx.act_freq_scores += (feature_acts.abs() > 0).float().sum(0)
```

Comment:
It seems like there's some inconsistenties in how sparsity is calculated in different parts of the code. Here, we use `acts.abs() > 0`, while for `did_fire`, we just use `acts > 0`. Also, we do `sum(0)` here, but `sum(-2)` for `did_fire`. I think these are all equivalent, since the acts have 2 dimensions, (so 0 == -2), and there's a ReLU in the autoencoder, so `feature_acts.abs() > 0` is the same as `feature_acts > 0` probably. Still, this feels like it could cause bugs in the future if we change the activation function of the autoencoder in the future. We should probably calculate this stuff the same way in both places.

Reply:
good catch. if this is tested, then setting them all to the same method + getting identical results on a small run seems like a good way to verify. Thanks for raising this.